joblib/numpy_pickle.py

```diff
@@ -1,229 +1,251 @@
+        We added some code to support versions of numpy prior to 1.9.
+        """
+        # Set buffer size to 16 MiB to hide the Python loop overhead.
+        buffersize = max(16 * 1024 ** 2 // array.itemsize, 1)
+        np_ver = [int(x) for x in pickler.np.__version__.split('.', 2)[:2]]
+
+        if array.dtype.hasobject:
+            # We contain Python objects so we cannot write out the data
+            # directly. Instead, we will pickle it out with version 2 of the
+            # pickle protocol.
+            pickle.dump(array, pickler.file, protocol=2)
+        elif self.order == 'F':
+            if pickler.np.compat.isfileobj(pickler.file):
+                array.T.tofile(pickler.file)
+            else:
+                # 'tobytes' is available since numpy 1.9.
+                if np_ver[0] >= 1 and np_ver[1] >= 9:
+                    for chunk in pickler.np.nditer(array,
+                                                   flags=['external_loop',
+                                                          'buffered',
+                                                          'zerosize_ok'],
+                                                   buffersize=buffersize,
+                                                   order='F'):
+                        pickler.file.write(chunk.tobytes('C'))
+                else:
+                    pickler.file.write(array.T.tostring('C'))
+        else:
+            if pickler.np.compat.isfileobj(pickler.file):
+                array.tofile(pickler.file)
+            else:
+                # 'tobytes' is available since numpy 1.9.
+                if np_ver[0] >= 1 and np_ver[1] >= 9:
+                    for chunk in pickler.np.nditer(array,
+                                                   flags=['external_loop',
+                                                          'buffered',
+                                                          'zerosize_ok'],
+                                                   buffersize=buffersize,
+                                                   order='C'):
+                        pickler.file.write(chunk.tobytes('C'))
+                else:
+                    pickler.file.write(array.tostring('C'))
+
+    def read_array(self, unpickler):
+        """Read array from unpickler file handle.
+
+        This function is an adapation of the numpy read_array function
+        available in version 0.10 in numpy/lib/format.py.
+        """
+        if len(self.shape) == 0:
+            count = 1
+        else:
+            count = unpickler.np.multiply.reduce(self.shape)
+        # Now read the actual data.
+        if self.dtype.hasobject:
+            # The array contained Python objects. We need to unpickle the data.
+            array = pickle.load(unpickler.file_handle)
+        else:
+            if (unpickler.np.compat.isfileobj(unpickler.file_handle) and
+                    count*self.dtype.itemsize/1e6 > 700):
```

Comment:
> I would worry too much about an overhead of 30MB

I wouldn't worry either, but basically my little code snippet doesn't reproduce what @aabadie was seeing so that is a bit weird.